storage/modules/block_storage.go

```diff
@@ -1084,6 +1184,108 @@ func (b *BlockStorage) FindTransaction(
 	return newestBlock, newestTransaction, nil
 }
 
+func (b *BlockStorage) FindRelatedTransactions(
+	ctx context.Context,
+	transactionIdentifier *types.TransactionIdentifier,
+	db database.Transaction,
+) (*types.BlockIdentifier, *types.Transaction, []*types.Transaction, error) {
+	rootBlock, tx, err := b.FindTransaction(ctx, transactionIdentifier, db)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	if rootBlock == nil {
+		return nil, nil, nil, nil
+	}
+
+	childIds, err := b.getChildren(ctx, tx, db)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// create map of seen transactions to avoid duplicates
+	seen := make(map[string]struct{})
+	children := []*types.Transaction{}
+
+	i := 0
+	for {
+		if i >= len(childIds) {
+			break
+		}
+		childID := childIds[i]
+		i++
+
+		// skip duplicates
+		if _, ok := seen[childID.Hash]; !ok {
+			seen[childID.Hash] = struct{}{}
+		} else {
+			continue
+		}
+
+		childBlock, childTx, err := b.FindTransaction(ctx, childID, db)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+
+		if childBlock == nil {
+			return nil, nil, nil, nil
+		}
+
+		children = append(children, childTx)
+		if rootBlock.Index < childBlock.Index {
+			rootBlock = childBlock
+		}
+
+		newChildren, err := b.getChildren(ctx, childTx, db)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		childIds = append(childIds, newChildren...)
+	}
+
+	return rootBlock, tx, children, nil
+}
+
+// TODO: add support for relations across multiple networks
+func (b *BlockStorage) getChildren(
+	ctx context.Context,
+	tx *types.Transaction,
+	db database.Transaction,
+) ([]*types.TransactionIdentifier, error) {
+	var children []*types.TransactionIdentifier
+	for _, relatedTx := range tx.RelatedTransactions {
+		// skip if on another network
+		if relatedTx.NetworkIdentifier != nil {
+			continue
+		}
+
+		if relatedTx.Direction == types.Forward {
+			children = append(children, relatedTx.TransactionIdentifier)
+		}
+	}
+
+	_, err := db.Scan(
+		ctx,
+		getBackwardRelationKey(tx.TransactionIdentifier, nil),
+		getBackwardRelationKey(tx.TransactionIdentifier, nil),
+		func(k []byte, v []byte) error {
+			ss := strings.Split(string(k), "/")
+			txHash := ss[len(ss)-1]
+			txID := &types.TransactionIdentifier{Hash: txHash}
+			children = append(children, txID)
+			return nil
+		},
+		false,
+		false,
+	)
```

Comment:
Is the purpose of this scan to find transactions where `tx` appears as a backward relation? If so, a comment here would be great! Sorry about dumb asks, this stuff is super confusing to me at first glance 😄

Reply:
no worries definitely a good ask! not a very readable line.